### ISSUE TITLE
[webview_flutter] camera not opening for (input type="file" capture="user" accept="image/*")

### ISSUE THREAD

**Issue by Shubhankt1** [flutter/flutter#97127]
If I click on button with HTML tag `<input type="file" capture="user" accept="image/*">`, the file chooser activity starts on Android while camera opens directly on iOS (and in browser on a PC/Mac).

FYI: I had made changes in the package to accomodate file upload [issue](https://github.com/flutter/flutter/issues/27924) from [HERE](https://github.com/flutter/plugins/pull/3225#issuecomment-831058022)

## Steps to Reproduce

1. Run simple index.html with tag `<input type="file" capture="user" accept="image/*">` in webview_flutter
2. Try to click camera (capture)

**Expected results:**  Directly camera intent should open to click the image

**Actual results:**  File chooser intent opens up (Android) where again I have to click the camera option.

**Comment by Shubhankt1** [comment]
@tneotia @fesp could you guys please help? I really need a way out for this.

**Comment by darshankawar** [comment]
@Shubhankt1 
If I understand your case correctly, you are adding an input type="file"  in `webview_flutter` plugin example code and then trying to run it on Android. What are the next steps you are carrying out ? By any chance can you check how is the behavior in native Android with the same case ? ie, whether it open file chooser intent or directly the camera intent ?

Can you also check [this](https://stackoverflow.com/questions/60289932/is-there-a-way-of-making-a-flutter-webview-use-android-camera-for-file-upload-h) SO link and see if it helps ?

**Comment by Shubhankt1** [comment]
@darshankawar 
Yes that's what i'm doing. I have input type="file" capture="user" tag and when I click on it, it opens the file chooser intent whereas expected behaviour is that it should open the camera directly.

I'm sorry, I dont know how to check it in native java, although I'm trying to fix this in the webview_flutter package code (java) itself.
I was thinking maybe in the `onShowFileChooser()` method somewhere I could check if it is requesting camera directly and then do appropriate logic. just a thought.

P.S.: The link you provided is what i did few months back to enable file chooser within webview. but it doesnt help when camera needs to open directly.

**Comment by darshankawar** [comment]
@Shubhankt1 
Can you try below options and see if either of them helps ?

`<input type="file" name="file" accept="image/*" capture>` or `<input type="file" name="file" accept="image/*" capture="camera">`

**Comment by Shubhankt1** [comment]
Hi @darshankawar 
just FYI, the same code which I have works fine in iOS but not in android.

I'm trying with the tags you sent. will update you soon on it.

**Comment by Shubhankt1** [comment]
Hi @darshankawar 
I tried those tags and they're not working.

in Android it shows the intent to choose between camera and files rather than opening camera directly.
in iOS it shows the option for choosing image from gallery or from files. no camera option at all.

**Comment by maheshmnj** [comment]
@Shubhankt1 when your add it natively in an HTML file does that launch camera in your phone?

```
<input type="file" name="file" accept="image/*" capture> or <input type="file" name="file" accept="image/*" capture="camera">
```

**Comment by Shubhankt1** [comment]
Hi @maheshmnj I didnt understand your question completely.

But I'm guessing you're trying to ask if camera got opened directly when I loaded webview using html file with those tags, right?
I tried that and this is the outcome,

> Hi @darshankawar I tried those tags and they're not working.
> 
> in Android it shows the intent to choose between camera and files rather than opening camera directly. in iOS it shows the option for choosing image from gallery or from files. no camera option at all.

**Comment by maheshmnj** [comment]
@Shubhankt1 I meant when you add the above input tag in a normal HTML file (No Flutter) and serve the web page locally and launch the web page on phone do you see the camera app launching?

**Comment by Shubhankt1** [comment]
Hi @maheshmnj 
I tried with [THIS LINK](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/capture)

When I opened this in android browser directly, it opened the camera directly, but if I opened it in the `webview_flutter` WebView, it opened File chooser intent.

Also, FYI, for now, I made some changes in the pckage's java code and instead of opening FileChooserIntet via `startActivityForResult(captureImageIntent, FILE_CHOOSER_REQUEST_CODE);` instead of using `startActivityForResult(captureImageIntent, FILE_CHOOSER_REQUEST_CODE);`. After this I cud successfully open camera directly but now file chooser doesnt open at all. Could you suggest some way to differentiate both? so that I can open camera activity ONLY when required?

Thanks!

**Comment by maheshmnj** [comment]
Thanks for the info @Shubhankt1,  choose inputType file is currently not supported (including accept="image"), I believe this is blocked by  https://github.com/flutter/flutter/issues/97127, I verified using the below code sample the camera opens fine on native web but none of the input types respond in flutter webview.

Native web | Flutter webview
:-: | :-:
<video src='https://user-images.githubusercontent.com/31410839/151920365-0b8bef05-9d5e-48ba-bb99-277c7157e0fc.mov' width=180/> | <video src='https://user-images.githubusercontent.com/31410839/151920369-fc00c6e5-735d-4c65-9265-c2bb4c225ecb.mov' width=180/>



<details>
<summary>code sample</summary>

#### assets/sample.html

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Flutter file input</title>
    <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700" rel="stylesheet">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
  </head>
  <body>
    <p>
        <label for="soundFile">What does your voice sound like?:</label>
        <input type="file" id="soundFile" capture="user" accept="audio/*">
        </p>
        <p>
        <label for="videoFile">Upload a video:</label>
        <input type="file" id="videoFile" capture="environment" accept="video/*">
        </p>
        <p>
        <label for="imageFile">Upload a photo of yourself:</label>
        <input type="file" id="imageFile" capture="user" accept="image/*">
        </p>
  </body>
</html>
```
### main.dart
```dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        color: Colors.white,
        title: 'Flutter Demo',
        theme: ThemeData(primarySwatch: Colors.blue),
        home: const WebViewSample());
  }
}

class WebViewSample extends StatefulWidget {
  const WebViewSample({Key? key}) : super(key: key);

  @override
  _WebViewSampleState createState() => _WebViewSampleState();
}

class _WebViewSampleState extends State<WebViewSample> {
  @override
  void initState() {
    super.initState();
    // Enable hybrid composition.
    if (Platform.isAndroid) WebView.platform = SurfaceAndroidWebView();
  }

  late WebViewController _controller;

  _loadHtmlFromAssets() async {
    String fileText = await rootBundle.loadString('assets/sample.html');
    _controller.loadUrl(Uri.dataFromString(fileText,
            mimeType: 'text/html', encoding: Encoding.getByName('utf-8'))
        .toString());
  }

  @override
  Widget build(BuildContext context) {
    final size = MediaQuery.of(context).size;
    return Container(
        height: size.height,
        width: size.width,
        child: WebView(
          initialUrl: '',
          javascriptMode: JavascriptMode.unrestricted,
          onWebViewCreated: (WebViewController webViewController) async {
            _controller = webViewController;
            _loadHtmlFromAssets();
          },
          onPageFinished: (value) {},
        ));
  }
}

```
</details>

This issue could be a possible duplicate of the linked issue if the team confirm this will be solved by #97127

cc: @ditman

**Comment by Shubhankt1** [comment]
@maheshmnj @ditman @darshankawar @fesp I recently updated `webview_flutter` from `2.0.4` to `3.0.1` and now file upload again stopped working.
when I tried to use the solution mentioned in [this PR](https://github.com/flutter/plugins/pull/3225), I found that the directory structure of the plugin is changed the files are different as well.
Could you let me know how to move ahead?

I need to do a release for my app and now I'm stuck... :/

**Comment by ditman** [comment]
> I need to do a release for my app and now I'm stuck... :/

I'd recommend you downgrade to a known working version. Webview had a very recent big refactor to enable it to work in more platforms (the change from 2 to 3), so the patch that you're referencing is probably not directly applicable any longer.

**Comment by Shubhankt1** [comment]
I tried downgrading back to `webview_flutter:^2.0.4` but it still doesn't give me the older directory structure.

**Comment by ditman** [comment]
@Shubhankt1 you can clone the repo into the hash you need, and then make an override to the `webview_flutter` package at that point. As I said, the webview has undergone a big rewrite recently, so it's likely that old recipes broke. The best way to go about this would be to update the fix and send a new PR.

**Comment by omerx** [comment]
Currently, it doesn't open the camera directly like in Chrome? Is there a solution that I can fix manually?


An app I bought is using it. I can't get support from the developer.

I guess it won't help if I update the version. The version I am using:   webview_flutter: ^2.0.8


**Steps to Reproduce**
Run simple index.html with tag <input type="file" capture="user" accept="image/*"> in webview_flutter
Try to click camera (capture)
Expected results: Directly camera intent should open to click the image

Actual results: File chooser intent opens up (Android) where again I have to click the camera option.


![image](https://user-images.githubusercontent.com/1086664/160924132-f3c8e5aa-0ef8-49e6-8138-fe1624c1a1b0.png)

**Comment by bparrishMines** [comment]
Fixed by https://github.com/flutter/plugins/pull/6881

**Comment by ditman** [comment]
Good one @bparrishMines !!
